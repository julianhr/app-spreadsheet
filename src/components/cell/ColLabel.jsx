/* @jsx jsx */
import React from 'react'
import PropTypes from 'prop-types'
import styled from '@emotion/styled'
import { jsx, css } from '@emotion/core' // eslint-disable-line
import { connect } from 'react-redux'

import { setColWidthDelta } from '~/actions/tableMetaActions'
import { DEFAULT_COL_WIDTH, MIN_COL_WIDTH } from '~/library/constants'
import { debounce } from '~/library/utils'


let InnerBorder = styled.div`
  position: relative;
  display: flex;
  align-items: center;
  justify-content: center;
  border-right: 1px solid ${props => props.theme.colors.cell.borderDark};
  border-bottom: 1px solid ${props => props.theme.colors.cell.borderDark};
  background: ${props => props.theme.colors.cell.labelBkg};
  font-size: 12px;
`

const BtnWidth = styled.button`
  background: transparent;
  cursor: col-resize;
  outline: none;
  position: absolute;
  border: 3px solid transparent;
  height: 100%;
  margin: 0 -4px 0 0;
  padding: 1px;
  right: 0;
  top: 0;
  z-index: 1;
`

class ColLabel extends React.PureComponent {

  static propTypes = {
    label: PropTypes.string.isRequired,
    setColWidthDelta: PropTypes.func.isRequired,
    width: PropTypes.number.isRequired,
  }

  static defaultProps = {
    width: DEFAULT_COL_WIDTH,
  }

  constructor() {
    super()
    this.handleBtnWidthOnMouseDown = this.handleBtnWidthOnMouseDown.bind(this)
    this.handleDocumentMouseMove = this.handleDocumentMouseMove.bind(this)
    this.handleDocumentMouseUp = this.handleDocumentMouseUp.bind(this)
    this.debSetColWidth = this.debSetColWidth.bind(this)
  }

  state = {
    pageX: 0,
    width: 0,
    btnRight: 0,
    isResizing: false,
  }

  getBtnWidthStyle() {
    if (!this.state.isResizing) { return }

    const newRight = this.state.btnRight + 4
    const minRight = this.state.width - MIN_COL_WIDTH
    const right = Math.min(newRight, minRight)

    return (theme) => css`
      right: ${right}px;
      margin-top: -70px;
      height: 95vh;
      border: 0;
      background: ${theme.colors.primary.live};
    `
  }

  debSetColWidth = debounce(40, (event) => {
    const delta = this.state.pageX - event.pageX
    this.setState({ btnRight: delta, })
  })

  handleDocumentMouseMove(event) {
    this.debSetColWidth(event)
  }

  handleDocumentMouseUp(event) {
    const delta = event.pageX - this.state.pageX
    const width = this.state.width + delta

    this.props.setColWidthDelta(this.props.label, width)
    this.setState({ btnRight: 0, isResizing: false })
    document.removeEventListener('mousemove', this.handleDocumentMouseMove)
    document.removeEventListener('mouseup', this.handleDocumentMouseUp)
  }

  handleBtnWidthOnMouseDown(event) {
    event.preventDefault()

    this.setState({
      pageX: event.pageX,
      width: this.props.width,
      isResizing: true,
    })

    document.addEventListener('mousemove', this.handleDocumentMouseMove)
    document.addEventListener('mouseup', this.handleDocumentMouseUp)
  }
  
  // handleBtnWidthOnMouseMove(event) {
  //   if (!this.state.isSettingWidth) { return }
  //   document.addEventListener('mousemove', (event) => {

  //   })
  //   const { pageX, pageY } = event
  //   console.log('handle', pageX, pageY)
  // }

  render() {
    const { label, width } = this.props

    return (
      <InnerBorder
        data-col={label}
        className='col-label-height'
        css={css`
          width: ${width}px;
        `}
      >
        {label}
        <BtnWidth
          css={this.getBtnWidthStyle()}
          tabIndex='-1'
          onMouseDown={this.handleBtnWidthOnMouseDown}
          onMouseMove={this.handleBtnWidthOnMouseMove}
          onMouseUp={this.handleBtnWidthOnMouseUp}
        />
      </InnerBorder>
    )
  }
}

function mapStateToProps(state, ownProps) {
  const width = state.tableMeta.colWidths[ownProps.label]
  return { width }
}

const mapDispatchToProps = { setColWidthDelta }

export default connect(mapStateToProps, mapDispatchToProps)(ColLabel)
