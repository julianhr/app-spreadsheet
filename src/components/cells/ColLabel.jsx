import React from 'react'
import PropTypes from 'prop-types'
import styled from '@emotion/styled'


const InnerBorder = styled.div`
  display: flex;
  align-items: center;
  justify-content: center;
  border-right: 1px solid ${props => props.theme.colors.cell.borderDark};
  border-bottom: 1px solid ${props => props.theme.colors.cell.borderDark};
  background: ${props => props.theme.colors.cell.labelBkg};
  font-size: 12px;
`

function ColLabel({ label }) {
  return (
    <td>
      <InnerBorder
        className='col-label-width col-label-height'
      >
        {label}
      </InnerBorder>
    </td>
  )
}

ColLabel.propTypes = {
  label: PropTypes.string,
}

export default ColLabel
