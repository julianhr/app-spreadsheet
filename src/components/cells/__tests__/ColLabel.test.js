import React from 'react'
import { render, fireEvent, cleanup } from 'react-testing-library'

import MockApp from '~/__tests__/__mocks__/MockApp'
import ColLabel from '../ColLabel'


const renderApp = (props) => {
  const wrapper = render(
    <MockApp>
      <table>
        <tr>
          <ColLabel {...props} />
        </tr>
      </table>
    </MockApp>
  )

  return wrapper
}

describe('ColLabel', () => {
  afterEach(cleanup)

  it('matches snapshot', () => {
    const props = {
      label: 'A'
    }

    renderApp(props)
    const el = document.querySelector('td')
    expect(el).toMatchSnapshot()
  })
})
