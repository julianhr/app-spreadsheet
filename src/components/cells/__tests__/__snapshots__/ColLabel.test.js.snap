// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ColLabel matches snapshot 1`] = `
<td>
  <div
    class="col-label-width col-label-height css-19eh4ji-InnerBorder e1al44mo0"
  >
    A
  </div>
</td>
`;
